src/HttpRequestHandler.cpp

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2021 Pei-Ru Wang, Cheryl Huang, Yuan Zhou. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+/**
+ * Project MyHttpServer
+ * @author Pei-Ru Wang, Cheryl Huang, Yuan Zhou
+ * @version 0.0.1
+ */
+
+#include "HttpRequestHandler.h"
+
+#include "HttpRouterHandler.h"
+
+/**
+ * HttpRequestHandler implementation
+ */
+
+HttpRequestHandler::HttpRequestHandler(
+    boost::asio::io_context *io_context,
+    std::unordered_map<int, handler_ptr> *map_handler_list, const int n_count)
+    : map_handler_list(*map_handler_list),
+      n_count(n_count),
+      strand(*io_context),
+      HttpHandler(io_context) {}
+
+HttpRequestHandler::~HttpRequestHandler() {
+  HttpRequestHandler::Stop();
+  std::cout << n_count << ":~HandlerComplete" << std::endl;
+}
+
+void HttpRequestHandler::Start() {
+  auto self(shared_from_this());
+  map_handler_list.insert(std::make_pair(n_count, self));
+
+  HttpRequestHandler::HandleRead();
+}
+
+void HttpRequestHandler::Stop() {
+  auto it = map_handler_list.find(n_count);
+  map_handler_list.erase(it);
+  socket.shutdown(boost::asio::ip::tcp::socket::shutdown_both);
+  socket.close();
+}
+
+void HttpRequestHandler::HandleRead() {
+  auto self(shared_from_this());
+  socket.async_read_some(
```

Comment:
Doing async_read_some() once might not yield all the data.

Reply:
Either add a TODO on handling such a case, or read it multiple times until you get the entire request.

Reply:
If it is not finished(result != success or fail), it will repeat HandleRead(), see line 66.